connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListenerBus.scala

```diff
@@ -0,0 +1,155 @@
+      try {
+        sparkSession.execute(cmdBuilder.build())
+      } catch {
+        case e: Exception =>
+          logWarning("Failed to remove the listener, please remove it again.", e)
+          return
+      }
+      if (executionThread.isDefined) {
+        executionThread.get.interrupt()
+        executionThread = Option.empty
+      }
+    }
+    listeners.remove(listener)
+  }
+
+  def list(): Array[StreamingQueryListener] = lock.synchronized {
+    listeners.asScala.toArray
+  }
+
+  def registerServerSideListener(): CloseableIterator[ExecutePlanResponse] = {
+    val cmdBuilder = Command.newBuilder()
+    cmdBuilder.getStreamingQueryListenerBusCommandBuilder
+      .setAddListenerBusListener(true)
+
+    val plan = Plan.newBuilder().setCommand(cmdBuilder.build()).build()
+    val iterator = sparkSession.client.execute(plan)
+    while (iterator.hasNext) {
+      val response = iterator.next()
+      if (response.getStreamingQueryListenerEventsResult.hasListenerBusListenerAdded &&
+        response.getStreamingQueryListenerEventsResult.getListenerBusListenerAdded) {
+        return iterator
+      }
+    }
+    iterator
+  }
+
+  def queryEventHandler(iter: CloseableIterator[ExecutePlanResponse]): Unit = {
+    try {
+      while (iter.hasNext) {
+        val response = iter.next()
+        val listenerEvents = response.getStreamingQueryListenerEventsResult.getEventsList
+        listenerEvents.forEach(event => {
+          event.getEventType match {
+            case StreamingQueryEventType.QUERY_PROGRESS_EVENT =>
+              postToAll(QueryProgressEvent.fromJson(event.getEventJson))
+            case StreamingQueryEventType.QUERY_IDLE_EVENT =>
+              postToAll(QueryIdleEvent.fromJson(event.getEventJson))
+            case StreamingQueryEventType.QUERY_TERMINATED_EVENT =>
+              postToAll(QueryTerminatedEvent.fromJson(event.getEventJson))
+            case _ =>
+              logWarning(s"Unknown StreamingQueryListener event: $event")
+          }
+        })
+      }
+    } catch {
+      case e: Exception =>
+        logWarning("Failed to handle the event, please add the listener again.", e)
+        lock.synchronized {
+          executionThread = Option.empty
+          listeners.forEach(remove(_))
+        }
+    }
+
+  }
+
+  def postToAll(event: Event): Unit = lock.synchronized {
+    listeners.forEach(
+      listener => try {event match {
+        case t: QueryStartedEvent =>
+          listener.onQueryStarted(event.asInstanceOf[QueryStartedEvent])
+        case t: QueryProgressEvent =>
+          listener.onQueryProgress(event.asInstanceOf[QueryProgressEvent])
+        case t: QueryIdleEvent =>
+          listener.onQueryIdle(event.asInstanceOf[QueryIdleEvent])
+        case t: QueryTerminatedEvent =>
+          listener.onQueryTerminated(event.asInstanceOf[QueryTerminatedEvent])
+        case _ => logWarning(s"Unknown StreamingQueryListener event: $event")
+      }
+      } catch {
```

Comment:
nit: this format looks a little bit unintuitive (two curly brackets at the same position in two lines...) , mind formatting it?

Reply:
sure, will update that. Might be some formatting issue when I copied the code...